runtime/jcl/common/jdk_internal_misc_ScopedMemoryAccess.cpp

```diff
@@ -84,27 +120,79 @@ Java_jdk_internal_misc_ScopedMemoryAccess_closeScope0(JNIEnv *env, jobject insta
 	} else {
 		J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
 		while (NULL != walkThread) {
+			jboolean threadHasScope = JNI_FALSE;
+			jint topStackFrameType = 0;
 			J9StackWalkState walkState;
 			walkState.walkThread = walkThread;
 			walkState.flags = J9_STACKWALK_ITERATE_FRAMES | J9_STACKWALK_ITERATE_O_SLOTS;
 			walkState.skipCount = 0;
 			walkState.userData1 = (void *)scope;
-			walkState.userData2 = (void *)&scopeNotFound;
+			walkState.userData2 = (void *)&threadHasScope;
+			walkState.userData3 = (void *)&topStackFrameType;
 			walkState.frameWalkFunction = closeScope0FrameWalkFunction;
 			walkState.objectSlotWalkFunction = closeScope0OSlotWalkFunction;
 
+			fprintf(stderr, "Thread: %s\n", (char *)walkThread->omrVMThread->threadName);
+
 			vm->walkStackFrames(walkThread, &walkState);
-			if (JNI_FALSE == *(jboolean *)walkState.userData2) {
+			if (JNI_TRUE == *(jboolean *)walkState.userData2) {
 				/* scope found */
-				break;
+				scopeNotFound = JNI_FALSE;
+				if (1 == *(jint *)walkState.userData3) {
+					/* Only set the exception if the top frame is interpreted.
+					 * If the top frame is JITted, the exception won't be caught.
+					 */
+					fprintf(stderr, "    Top is interp\n");
+					if (NULL == head) {
+						current = (LinkedThreads *)j9mem_allocate_memory(sizeof(LinkedThreads), J9MEM_CATEGORY_VM);
+						if (NULL != current) {
+							current->thread = walkThread;
+							current->next = NULL;
+							head = current;
+							fprintf(stderr, "    First LL node\n");
+						}
+					} else {
+						current->next = (LinkedThreads *)j9mem_allocate_memory(sizeof(LinkedThreads), J9MEM_CATEGORY_VM);
+						if (NULL != current->next) {
+							current->next->thread = walkThread;
+							current->next->next = NULL;
+							current = current->next;
+							fprintf(stderr, "    Next LL node\n");
+						}
+					}
+				}
 			}
 
 			walkThread = J9_LINKED_LIST_NEXT_DO(vm->mainThread, walkThread);
 		}
 	}
 
+	current = head;
+	while (NULL != current) {
+		/* Scope is being accessed by a thread */
+		J9VMThread *scopeThread = current->thread;
+		LinkedThreads *prev = NULL;
+
+		if (NULL == scopeThread->currentException) {
```

Comment:
`stopThrowable` will not be set if `currentException` is `null`. On PPCLE, the JIT is capable of setting exceptions for hardware events. The same applies to S390 (zOS/zLinux). Refer to `jitPPCHandler`: https://github.com/eclipse-openj9/openj9/blob/a2a4729e46530270e54277629bfb8252766e4bab/runtime/compiler/runtime/SignalHandler.c#L572-L574
This will set the `currentException` and prevent `stopThrowable` from being set in your code. Most probably, this is the cause for the intermittent `TestHandshake` timeout failures on PPCLE. For fixing these failures, you will need to clear `currentException` and always set `stopThrowable`.

@gacholio Can we set `currentException` to `null` for clearing it?

Reply:
currentException is only set from the current thread, and should only be read from the current thread. Re-using stopThrowable is going to cause problems interacting with stop(), which is deprecated but still hasn't been removed.